harness/determined/experimental/inference/_torch_batch_process.py

```diff
@@ -0,0 +1,210 @@
+import abc
+import logging
+import math
+import json
+import os
+import pathlib
+from dataclasses import dataclass
+
+import torch
+import torch.distributed as dist
+
+from torch.utils.data import Dataset
+from typing import Callable, Optional, Type
+
+import determined as det
+from determined import core
+from determined.common import set_logger
+from determined.pytorch import DataLoader
+from determined.tensorboard.util import get_rank_aware_path
+
+set_logger(False)
+
+
+@dataclass
+class TorchBatchInitInfo:
+    worker_rank: int
+    default_device: torch.device
+    tensorboard_path: str
+
+
+def get_default_device(core_context) -> torch.device:
+    local_rank = core_context.distributed.local_rank
+    local_num_gpu = torch.cuda.device_count()
+    if local_rank >= local_num_gpu:
+        return torch.device("cpu")
+    else:
+        return torch.device("cuda", local_rank)
+
+
+def initialize_distributed_backend() -> Optional[core.DistributedContext]:
+    info = det.get_cluster_info()
+
+    distributed_backend = det._DistributedBackend()
+    if distributed_backend.use_torch():
+        if torch.cuda.is_available():
+            dist.init_process_group(backend="nccl")  # type: ignore
+        else:
+            dist.init_process_group(backend="gloo")  # type: ignore
+        return core.DistributedContext.from_torch_distributed()
+    elif info and (len(info.container_addrs) > 1 or len(info.slot_ids) > 1):
+        raise ValueError(
+            "In multi-slot managed cluster training, you must wrap your training script with a "
+            "distributed launch layer such as determined.launch.torch_distributed"
+        )
+    return None
+
+
+def initialize_default_inference_context() -> core.Context:
+    distributed_context = initialize_distributed_backend()
+    # Setting preempt mode to WorkerAskMaster makes the call non-blocking
+    # We are also ok if workers are preempted at different batch idx
+    return det.core.init(
+        distributed=distributed_context, preempt_mode=core.PreemptMode.WorkersAskMaster
+    )
+
+
+class TorchBatchProcessor(metaclass=abc.ABCMeta):
+    """
+    User can initialize necessary resources in the init function, such as
+    - model for prediction
+    - storage client (e.g. s3 client)
+    """
+
+    @abc.abstractmethod
+    def __init__(self, core_context: core.Context, init_info: TorchBatchInitInfo) -> None:
```

Comment:
what if we made `TorchBatchInitInfo` a context object of sorts, where `get_default_device` could be a method in that context

Reply:
Not sure I understand why we may want to make it a context object? Could you please elaborate a bit? Thanks!

default_device is not expected to change over the life time of the object.

Reply:
I made the suggestion because you were passing in the core context as well as the `init_info`, and I was wondering if they could be combined in functionality to create a `TorchBatchProcessorContext` sort of object.

I see that you removed the core context, but I still wonder if we should expand the `init_info` data class to just a class. It will be the same from the user's perspective, but allows us more flexibility in terms of what we can support now or in the future. I mentioned in other comments some methods we could potentially support, but even if we don't, I feel a pure data class may limit us in the future. It could also make the code more organized in terms of which methods belong to which objects. Maybe something like:
```
class TorchBatchProcessorContext:
    def __init__(self, core_context: determined.core.Context):
        self.distributed = core_context.distributed
        self.device = self._init_device()

    def to_device(self):
        pass
    
    def get_tensorboard_path(self):
        pass
```

Reply:
Ahh got it, thanks for the detailed explanation. It makes sense to me, will update accordingly.